### PR TITLE
fix(3.0.3): Prompt in editor for 401 error, fix profile propagation

### LINKED ISSUE #3197: V3 doesn't prompt for credentials when opening file in editor
https://github.com/user-attachments/assets/68752325-5d53-4beb-9a60-b91c21c6f867

_Originally posted by @JillieBeanSim in https://github.com/zowe/zowe-explorer-vscode/pull/3157#pullrequestreview-2354573221_

### PR DESCRIPTION
**Pull request by traeok**
## Proposed changes

Port of #3302 for the 3.0.3 release.

Fixes #3197, #3289 

### How to test

- Checkout this branch in the repo with VS Code opened
- `pnpm install && pnpm build`
- After the build has finished, click on the Run & Debug panel on the VS Code Side Bar
- Select "Run VS Code Extension" from the tasks list at the top and select the Play ▶ button
- Once opened, make sure that you have (or set) valid credentials on a profile in ZE
- Execute a search on that profile and expand a folder so that a file/PDS member/spool item is visible
- Before opening the file, right click on your profile -> "Manage Profile" -> "Update Credentials"
- Enter in invalid credentials. I recommend using a fake username to prevent lockouts
- Once finished, try to open the file. Notice the error in the editor appears but you are prompted to update your credentials.
- Enter in the valid credentials
- Once done, click "Try again" in the editor to fetch & open the file

## Release Notes

Milestone: 3.0.3

Changelog:

- Fixed issue where users were not prompted to enter credentials if a 401 error was encountered when opening files, data sets or spools in the editor. [#3197](https://github.com/zowe/zowe-explorer-vscode/issues/3197)
- Fixed issue where profile credential updates or token changes were not reflected within the filesystem. [#3289](https://github.com/zowe/zowe-explorer-vscode/issues/3289)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

### CODE REVIEW

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
Hey @traeok thanks for porting these changes.  I do see some issues in this branch.  In MVS I don't get the pop up and in USS the file is holding outdated info with old credentials.  Using same testing scenario shared in original PR

**Review by zFernand0 — APPROVED** [review]
I'm not opposed to adding these as part of the 3.0.3 fixes, but I'd like to be respectful for those who've already completed their testing on the 3.0.3 PR 😋 

In either case, I'm happy with these changes!

Thanks for porting! 😋

**Review by zFernand0 — APPROVED** [review]
TL;DR: LGTM! 😋 
---

Apologies for going back and forth on having the `promptForAuthError` in the `stat` v. the `readFile` methods.

> In MVS I don't get the pop up and in USS the file is holding outdated info with old credentials. Using same testing scenario shared in original PR

As it stand, these changes make the behavior more consistent since I only see the above behavior in somewhat of an edge case. This scenario may have uncovered a long-standing issue that we've had for quite some time and it's just becoming more visible since Zowe Explorer is more reactive thanks to the file system provider implementation.

Here are some steps to reproduce the problem of not getting a pop-up on the MVS tree. These same steps can be used to reproduce the issue in USS.
1. Open a PDS member (and keep the PDS expanded)
2. Update credentials (via right-click menu) to a dummy/bad user
3. Open another member in the same PDS (notice the pop-up) and DO NOT update your creds
4. Open the same member from step one (notice no pop-up)

Long story short, there may be other areas for improvement related to 401 handling. However, I do believe this PR is an overall improvement on what we have out in the field today 😋
